### ISSUE TITLE
Log if a follower receives a proposal with POLRound == LockedRound and the locked value is different than the proposed one

### ISSUE THREAD

**Issue by angbrav** [cometbft/cometbft#1309]
## Problem Definition
The code assumes that when a follower receives a proposal with `POLRound == LockedRound`, the locked value is equal to the proposed one ([line](https://github.com/cometbft/cometbft/blob/main/consensus/state.go#L1435)). This follows from the 2/3 assumption (at least 2/3 of the voting power is under the control of correct validators), which guarantees that (*) a consensus round cannot generate two certificates (+2/3) for different values.

## Proposal
I propose to log when a follower receives a proposal with `POLRound == LockedRound` and its locked value is different from the proposed one. This may indicate that the 2/3 assumption does not hold anymore. It should be discussed whether more drastic measures should be taken, as violating Property (*) may break safety.

**Comment by angbrav** [comment]
> I assume that you are referring to this [line](https://github.com/cometbft/cometbft/blame/49127d27ef23e82f2e7cbc1dea9dd3b90db726c3/consensus/state.go#L1459) of the code, when we compared the POLRound, assume to be set (> -1), to the locked round:

Yes

**Comment by cason** [comment]
A point that has to be made is that to accept a value with `POLRound > -1` we have to receive, or have received, `2/3+` voting power equivalent `Prevote` messages from `POLRound` for the proposed value. In terms of implementation, a node that proposes, or forward a `Proposal` with such content, also makes it available the corresponding `Prevote` messages that enable the receiver to attest that in round `POLRound` the proposed value has been declared valid.

At the same time, if `LockedRound == X`, the node has necessarily received `2/3+` voting power equivalent `Prevote` messages for `LockedValue` at round `X`.

So, if it happens that a `Proposal` with `POLRound == LockedRound > -1` but with proposed value different `LockedValue` is received and accepted, the node will have in its vote set `2/3+` voting power equivalent votes for both its `LockedValue` and the value proposed in the current round. In order to have this two quorums from different values at the same round, we necessarily (*) need some validators to double vote. This situation is observed when adding the second (equivocating) vote from one of such validators, whic leads the node to produce an evidence of misbehavior.

In other words, this situation is already caught by the evidence module of CometBFT, as it requires validators to double voting.

(*) This is true provided that the voting power of Byzantine nodes is not a majority of the total voting power. The implementation assumes that this does not happen, in particular because a majority of Byzantine voting power would prevent the propagation of evidences of misbehaviour.

**Comment by andynog** [comment]
Hi @angbrav, does the answer provided by @cason clarifies things from a consensus perspective ? Also, I believe this issues was opened in order to add additional logging on a particular case. Is this still pertinent ? If so, could you suggest where in the code do you think this should be implemented ?

**Comment by angbrav** [comment]
Hi @andynog! I understand what @cason is saying: if the assumption holds, the described scenario cannot happen. What I am proposing is just to log in case the scenario happens as this would indicate that the +2/3 assumption does not hold. This way, if something goes tremendously wrong, we may be able to figure out the cause. But I do not have a strong opinion on this matter.

Regarding where to log it, it should be around [this line](https://github.com/cometbft/cometbft/blob/49127d27ef23e82f2e7cbc1dea9dd3b90db726c3/consensus/state.go#L1459)

**Comment by andynog** [comment]
Hi @angbrav, since it's just logging, would you like to try to submit a PR with a proposed change? We'd love to have more external contributors and we'd be happy to review the PR. Thanks !

**Comment by angbrav** [comment]
I'll give it a try

**Comment by cason** [comment]
This is an interesting situation to log indeed.

I would suggest replacing the `<=` by `<` [here](https://github.com/cometbft/cometbft/blob/49127d27ef23e82f2e7cbc1dea9dd3b90db726c3/consensus/state.go#L1459), so that it also matches the content of the debug log.

Then add after [this](https://github.com/cometbft/cometbft/blob/49127d27ef23e82f2e7cbc1dea9dd3b90db726c3/consensus/state.go#L1469) clause, when we already know that the proposal doesn't match the locked value, the log and action for the discussed scenario.

**Comment by cason** [comment]
Here  #2042. @angbrav, I've tried to request a review from you, but for some reason it couldn't find your handler.

**Comment by cason** [comment]
Stretch goal:

- [ ] Add a test unit covering this scenario

**Comment by adizere** [comment]
@angbrav @cason is this still relevant?

**Comment by cason** [comment]
We can re-word it into another issue.

The point is that this code is currently unreachable, from my attempts to come up with a test unit to cover it.

**Comment by cason** [comment]
I will close this issue, as we don't really have bandwidth to handle this corner case scenarios in test units.
